src/http/Stream.cc

```diff
@@ -428,8 +428,14 @@ Http::Stream::buildRangeHeader(HttpReply *rep)
         range_err = "no [parse-able] reply";
     else if ((rep->sline.status() != Http::scOkay) && (rep->sline.status() != Http::scPartialContent))
         range_err = "wrong status code";
-    else if (hdr->has(Http::HdrType::CONTENT_RANGE))
-        range_err = "origin server does ranges";
+    else if (hdr->has(Http::HdrType::CONTENT_RANGE)) {
```

Comment:
It would be nice to clarify why we are using raw `HdrType::CONTENT_RANGE` instead of `rep->content_range` in this context.

Also, should we protect the fragile code further below from multipart 206 responses to single-part Range requests? If yes, then perhaps we can accomplish the above clarification and add this new protection using the following sketch?

```
else if (rep->sline.status() == Http::scPartialContent))
    range_err = "too complex response"; // probably contains what the client needs
else if (rep->sline.status() != Http::scOkay))
    range_err = "wrong status code";
else if (rep->content_range)
    Must(!"meaningless Content-Range escaped");
else if (hdr->has(Http::HdrType::CONTENT_RANGE))
    range_err = "meaningless response";
```

Do you know whether this method is reached when an unpatched Squid receives a multipart 206 response (without a Content-Range header) while the client was sending a single-part Range request? If it is reached, what protects the client from Squid extracting random bytes from that complex mutlipart response? Another check somewhere below?

If we are fixing the multipart handling bug, the PR description would need to be updated.

Reply:
> else if (rep->content_range)
>      Must(!"meaningless Content-Range escaped");

I am not sure we should care about 'content_range' field consistency here: another use of this field would require a similar check. It is better to create a 'getter' method for 'content_range' and put the check there:

```
HttpHdrContRange *
HttpReply::contentRange() const
{
    if (content_range)
        assert(sline.status() == Http::scPartialContent);
    return content_range;
}
```

Reply:
>Do you know whether this method is reached when an unpatched Squid receives a multipart 206 response (without a Content-Range header) while the client was sending a single-part Range request?

Yes, it is reached with 'range_offset_limit none' Squid option.

> what protects the client from Squid extracting random bytes from that complex mutlipart response?

Nothing protects, leading to a malformed response, sending back to the client. Your
'if (rep->sline.status() == Http::scPartialContent)' check should resolve this problem.

Reply:
> 4. Similarly, let's replace this raw usage with content_range if possible:
> ```
> if (!hdr->has(Http::HdrType::CONTENT_RANGE)) {
>    /* No content range, so this was a full object we are
>     * sending parts of.
>     */
> ```

There is no sense in doing this: after our checks have passed, we deal with only 200 responses and content_range==null for it.

Reply:
> I am not sure we should care about 'content_range' field consistency here: another use of this field would require a similar check.

That proposed code is _not_ about checking `content_range` consistency but about documenting why we are accessing raw `CONTENT_RANGE ` header value (while increasing our chances of keeping things consistent across future changes). This is an unusual place/code/usage. There is no similar place/code/usage right now AFAIK. If similar code appears, we may decide to add a wrapper method to avoid code duplication, but it is too early to know what what future method should look like.

> Your `if (rep->sline.status() == Http::scPartialContent)` check should resolve this problem.

Glad that clarifying this code also exposed and fixed a yet another nasty bug! Please update the PR description to mention this fixed problem (when it is fixed).

> There is no sense in [replace this raw usage with content_range]

I disagree. The motivation for this replacement is to reduce the number of places where we access the raw header. Raw header access does not go through our new "Is this a meaningful header?" check. Thus, _each_ such raw access is a red flag that requires careful analysis and may require additional comments, accompanying `content_range` checks, or other tricks. The fewer such raw accesses we leave in the code, the less time we will spend thinking about these red flags long-term, and the lower are the chances that the problem will resurface.

We cannot replace all raw accesses -- some of them are legitimate/correct and must stay as is (for various reasons). I see no justification for keeping this particular raw access though. If I missed one, please discuss.

> after our checks have passed, we deal with only 200 responses and content_range==null for it.

Yes, I know, but removing raw access is motivated by a completely different/unrelated concern. You have already added a new TODO to deal with the 206 generation code (and its preconditions). We know that the combination of preconditions and the 206 generation code was and remains wrong, with many of the code branches unused/unreachable, but fixing that (complex) problem is outside this issue scope.

Reply:
Done.